File: acwing/算法提高课/第四章高级数据结构/1282.搜索关键词/1282.搜索关键词.py
N = 10000 + 10
M = 55

tr = [[0 for _ in range(26)] for _ in range(N * M)]
cnt = [0 for _ in range(N * M)]
q = [0 for _ in range(N * M)]
ne = [0 for _ in range(N * M)]
hh = 0
tt = - 1
idx = 0

t = int(input())


def insert(s):
    global idx
    p = 0
    for i in range(len(s)):
        u = ord(s[i]) - ord('a')
        if tr[p][u] == 0:
            idx += 1
            tr[p][u] = idx
        p = tr[p][u]
    cnt[p] += 1


def build():
    global hh, tt
    for i in range(26):
        if tr[0][i]:
            tt += 1
            q[tt] = tr[0][i]

    while hh <= tt:

        t = q[hh]
        hh += 1
        for i in range(26):
            j = ne[t]
            c = tr[t][i]
            if c == 0:
                continue
            while j and tr[j][i] == 0:
                j = ne[j]
            if tr[j][i]:
                j = tr[j][i]
            ne[c] = j
            tt += 1
            q[tt] = c


for _ in range(t):
    tr = [[0 for _ in range(26)] for _ in range(N * M)]
    cnt = [0 for _ in range(N * M)]
    ne = [0 for _ in range(N * M)]
    hh = 0
    tt = - 1
    idx = 0
    n = int(input())
    for _ in range(n):
        s = input()
        insert(s)

    build()

    s = input()
    j = 0
    res = 0
    for i in range(len(s)):
        u = ord(s[i]) - ord('a')

        while j and tr[j][u] == 0:
            j = ne[j]
        if tr[j][u]:
            j = tr[j][u]

        p = j
        while p:
            res += cnt[p]
            cnt[p] = 0
            p = ne[p]

    print(res)

File: acwing/算法提高课/第四章高级数据结构/1282.搜索关键词/1282.搜索关键词.cpp
#include <iostream>
#include <cstring>
#include <algorithm>
#include <cstdio>

using namespace std;

const int N = 1e4 + 10, M = 1e6 + 10, S = 55;

char s[M];
int tr[N * S][26], q[N * S], idx, ne[N * S], cnt[N * S], hh = 0, tt = -1;

void insert()
{
    int p = 0;
    for (int i = 0; s[i]; i++)
    {
        int u = s[i] - 'a';
        if (!tr[p][u])
            tr[p][u] = ++idx;
        p = tr[p][u];
    }
    cnt[p] += 1;
}

void build()
{
    for (int i = 0; i < 26; i++)
    {
        if (tr[0][i])
            q[++tt] = tr[0][i];
    }

    while (hh <= tt)
    {
        int t = q[hh++];

        for (int i = 0; i < 26; i++)
        {
            int c = tr[t][i];
            if (!c)
            {
                tr[t][i] = tr[ne[t]][i];
            }
            else
            {
                ne[c] = tr[ne[t]][i];
                q[++tt] = c;
            }
        }
    }
}

int main()
{
    int t;
    scanf("%d", &t);
    while (t--)
    {
        memset(tr, 0, sizeof tr);
        memset(cnt, 0, sizeof cnt);
        memset(ne, 0, sizeof ne);
        idx = 0;
        hh = 0, tt = -1;
        int n;
        scanf("%d", &n);
        for (int i = 0; i < n; i++)
        {
            scanf("%s", s);
            insert();
        }
        build();

        scanf("%s", s);
        int res = 0;
        for (int i = 0, j = 0; s[i]; i++)
        {
            int u = s[i] - 'a';

            j = tr[j][u];
            int p = j;
            while (p && cnt[p] != -1)
            {
                res += cnt[p];
                cnt[p] = -1;
                p = ne[p];
            }
        }
        printf("%d", res);
    }
}